pkg/api/ocplifecycle/codefreeze.go

```diff
@@ -0,0 +1,71 @@
+package ocplifecycle
+
+import (
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+const (
+	ocpProductName = "ocp"
+)
+
+type OCPCodeFreezeLifecycle struct {
+	c                   *Config
+	timelineOpts        TimelineOptions
+	timeline            Timeline
+	when                time.Time
+	IsCloseToCodeFreeze func(codeFreezeDate, now time.Time, howManyDaysBefore int) bool
+}
+
+func NewOCPCodeFreezeLifecycle(c *Config, when time.Time) *OCPCodeFreezeLifecycle {
+	tOpts := TimelineOptions{
+		OnlyEvents: sets.NewString([]string{
+			string(LifecycleEventCodeFreeze),
+		}...),
+	}
+	t := c.GetTimeline(ocpProductName, tOpts)
+	return &OCPCodeFreezeLifecycle{
+		c:                   c,
+		timelineOpts:        tOpts,
+		timeline:            t,
+		when:                when,
+		IsCloseToCodeFreeze: isCloseToCodeFreeze,
+	}
+}
+
+func (ol *OCPCodeFreezeLifecycle) IsBeforeCodeFreeze(howManyDaysBefore int) bool {
+	_, after := ol.timeline.DeterminePlaceInTime(ol.when)
+	if after.LifecyclePhase.Event != LifecycleEventCodeFreeze {
+		return false
+	}
+	return ol.IsCloseToCodeFreeze(after.LifecyclePhase.When.Time, ol.when, howManyDaysBefore)
+}
+
+func (ol *OCPCodeFreezeLifecycle) GetOCPVersions() (cur, future *MajorMinor, e error) {
+	before, after := ol.timeline.DeterminePlaceInTime(ol.when)
+	logrus.Debugf("parsing current OCP version %s", before.ProductVersion)
+	currentVersion, err := ParseMajorMinor(before.ProductVersion)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	logrus.Debugf("parsing future OCP version %s", after.ProductVersion)
+	futureVersion, err := ParseMajorMinor(after.ProductVersion)
+	if err != nil {
+		return nil, nil, err
+	}
+	return currentVersion, futureVersion, nil
+}
+
+func isCloseToCodeFreeze(codeFreezeDate, now time.Time, howManyDaysBefore int) bool {
```

Comment:
This function can call `time.Now()` itself instead of getting it as an argument, but that won't be testable. Either way, the argument shouldn't be named `now` since it is not representing the present time.